src/main/java/net/minecraftforge/common/ForgeTagHandler.java

```diff
@@ -246,166 +224,44 @@ public static void setCustomTagTypes(Set<ResourceLocation> customTagTypes)
-     */
-    public static TagContainer withNoCustom(TagContainer tagCollectionSupplier)
+    public static <T> TagCollection.NetworkPayload serializeToNetwork(ResourceLocation location, TagCollection<T> tags)
     {
-        ImmutableMap.Builder<ResourceLocation, TagCollection<?>> builder = ImmutableMap.builder();
-        for (ResourceLocation registryName : customTagTypeNames)
-        {
-            StaticTagHelper<?> tagRegistry = StaticTags.get(registryName);
-            if (tagRegistry != null)
-            {
-                builder.put(registryName, TagCollection.of(Collections.emptyMap()));
-            }
-        }
-        return withSpecificCustom(tagCollectionSupplier, builder.build());
-    }
+        IForgeRegistry<?> reg = RegistryManager.ACTIVE.getRegistry(location);
+        //This should not throw, as before this method is called the registry name is checked against the custom tag type names
+        //but it gives a better error message then a ClassCastException below. (in case something goes horribly wrong)
+        validateRegistrySupportsTags(reg);
 
-    /**
-     * Gets an {@link IForgeTagContainer} with specific custom tag types for testing if any tags are missing.
-     *
-     * @apiNote Internal
-     */
-    public static TagContainer withSpecificCustom(TagContainer tagCollectionSupplier, Map<ResourceLocation, TagCollection<?>> customTagTypes)
-    {
-        return new TagContainer(tagCollectionSupplier.collections)
+        Map<ResourceLocation, Tag<T>> tagMap = tags.getAllTags();
+        Map<ResourceLocation, IntList> idMap = Maps.newHashMapWithExpectedSize(tagMap.size());
+        tagMap.forEach((name, tag) ->
         {
-            @Override
-            public Map<ResourceLocation, TagCollection<?>> getCustomTagTypes()
+            List<T> values = tag.getValues();
+            IntList idList = new IntArrayList(values.size());
+            for(T t : values)
             {
-                return customTagTypes;
+                idList.add(RegistryManager.ACTIVE.getRegistry(location).getID(((IForgeRegistry<?>)t).getRegistryName()));
             }
-        };
-    }
 
-    /**
-     * Helper storage class for keeping track of various data for all custom tag types in the NetworkTagReader to make the code easier to read.
-     *
-     * @apiNote Internal: For use by NetworkTagManager
-     */
-    public static class TagCollectionReaderInfo
-    {
-
-        private final ResourceLocation tagType;
-        private final TagLoader<?> reader;
-        private final Map<ResourceLocation, Tag.Builder> tagBuilders;
-
-        private TagCollectionReaderInfo(ResourceLocation tagType, TagLoader<?> reader, Map<ResourceLocation, Tag.Builder> tagBuilders)
-        {
-            this.tagType = tagType;
-            this.reader = reader;
-            this.tagBuilders = tagBuilders;
-        }
+            idMap.put(name, idList);
+        });
+        return new TagCollection.NetworkPayload(idMap);
```

Comment:
Given this method is just used in one place, if we end up going with this way of doing it over using the backing vanilla registry, this should probably just return the idMap so that `TagCollection.NetworkPayload` constructor does not have to be AT'd to being visible/public. (Either way the AT should be able to be removed as if we do switch to the `TagCollection#serializeToNetwork` method then we won't need access to the constructor either)